### PR TITLE
RMRPA-20 When starting a release-candidate, no version is created in Jira

### PR DESCRIPTION
**Pull request by Frazerus**
[Ticket](https://re-motion.atlassian.net/browse/RMRPA-20)
I think this fits for you too @a-ctor, if not, please just relay to Mike :)

### CODE REVIEW

File: Core.UnitTests/Steps/AddFixVersionsForNewReleaseBranchStepTests.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
+#nullable enable
```
Comment (a-ctor) [inline Core.UnitTests/Steps/AddFixVersionsForNewReleaseBranchStepTests.cs:18]: All of the new files are missing a license header?

File: Core.UnitTests/Steps/AddFixVersionsForNewReleaseBranchStepTests.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    step.Execute(currentVersion, nextJiraVersion);
+
+    jiraVersionCreatorMock.Verify();
+    jiraIssueServiceMock.Verify();
+    jiraIssueServiceMock.Verify(_ => _.AddFixVersionToIssues(issues, rcVersionId), Times.Once);
+    jiraIssueServiceMock.Verify(_ => _.AddFixVersionToIssues(issues, nextJiraVersionId), Times.Once);
+    inputReaderMock.Verify(_ => _.ReadConfirmation(It.IsAny<bool>()),Times.Exactly(2));
+
+  }[Test]
+  public void Execute_WithNoConfirmationGiven_DoesNotAddAnyVersionToIssues()
```
Comment (a-ctor) [inline Core.UnitTests/Steps/AddFixVersionsForNewReleaseBranchStepTests.cs]:
```suggestion
  }

  [Test]
  public void Execute_WithNoConfirmationGiven_DoesNotAddAnyVersionToIssues()
```

File: Core/Steps/SubSteps/AddFixVersionsForNewReleaseBranchSubStep.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+using Remotion.ReleaseProcessAutomation.Jira.ServiceFacadeImplementations;
+using Remotion.ReleaseProcessAutomation.ReadInput;
+using Remotion.ReleaseProcessAutomation.SemanticVersioning;
+using Spectre.Console;
+
+namespace Remotion.ReleaseProcessAutomation.Steps.SubSteps;
+
+public interface IAddFixVersionsForNewReleaseBranchSubStep
+{
+  public void Execute (SemanticVersion currentVersion, SemanticVersion nextJiraVersion);
```
Comment (a-ctor) [inline Core/Steps/SubSteps/AddFixVersionsForNewReleaseBranchSubStep.cs]: Interface methods never have an access modifier as they are always `public`

```suggestion
  void Execute (SemanticVersion currentVersion, SemanticVersion nextJiraVersion);
```

File: Core/Steps/SubSteps/AddFixVersionsForNewReleaseBranchSubStep.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      IJiraIssueService jiraIssueService,
+      IJiraVersionCreator jiraVersionCreator)
+      : base(console, inputReader, jiraIssueService, jiraVersionCreator)
+  {
+  }
+
+  public void Execute (SemanticVersion currentVersion, SemanticVersion nextJiraVersion)
+  {
+    var issues = JiraIssueService.FindAllNonClosedIssues(currentVersion.ToString());
+    Console.WriteLine($"These are some of the issues that are open for the current version '{currentVersion}':");
```
Comment (a-ctor) [inline Core/Steps/SubSteps/AddFixVersionsForNewReleaseBranchSubStep.cs:29]: Unfortunate that this looks exactly like a "normal" `Console.WriteLine()` call so there is no easy guard against not using the `IAnsiConsole`

File: Core/Steps/SubSteps/AddMoveCreateFixVersionsSubStepBase.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+using System.Linq;
+using Remotion.ReleaseProcessAutomation.Jira;
+using Remotion.ReleaseProcessAutomation.Jira.ServiceFacadeImplementations;
+using Remotion.ReleaseProcessAutomation.ReadInput;
+using Remotion.ReleaseProcessAutomation.SemanticVersioning;
+using Spectre.Console;
+
+namespace Remotion.ReleaseProcessAutomation.Steps.SubSteps;
+
+public class AddMoveCreateFixVersionsSubStepBase
```
Comment (a-ctor) [inline Core/Steps/SubSteps/AddMoveCreateFixVersionsSubStepBase.cs]: If it is a base should we make it abstract?

File: Core/Steps/SubSteps/ReleaseVersionAndMoveIssuesSubStep.cs [hunk 6]
```diff
@@ -32,29 +28,26 @@ public class ReleaseVersionAndMoveIssuesSubStep
...
-    _jiraVersionCreator = jiraVersionCreator;
     _jiraVersionReleaser = jiraVersionReleaser;
   }
 
   public void Execute (SemanticVersion currentVersion, SemanticVersion nextVersion, bool squashUnreleased = false, bool movePreReleaseIssues = false)
   {
-    var currentVersionID = CreateVersion(currentVersion);
-    var nextVersionID = CreateVersion(nextVersion);
+    var currentVersionID = CreateVersion(currentVersion.ToString());
+    var nextVersionID = CreateVersion(nextVersion.ToString());
```
Comment (a-ctor) [inline Core/Steps/SubSteps/ReleaseVersionAndMoveIssuesSubStep.cs]: Maybe we should add an overload in the base class to do this conversion, allowing us to use the `SemanticVersion` without needing to worry how to correctly convert it.

File: Core/Steps/SubSteps/AddFixVersionsForNewReleaseBranchSubStep.cs (hunk 4 above)
Comment (Frazerus) [reply Core/Steps/SubSteps/AddFixVersionsForNewReleaseBranchSubStep.cs:29]: We might want to change all console fields to some better name, but I think that should be an extra ticket?

Comment (MichaelKetting) [reply Core/Steps/SubSteps/AddFixVersionsForNewReleaseBranchSubStep.cs:29]: yes, extra ticket. And I agree with @a-ctor  when reviewing without highlighting. I have no good naming solutions, fortunately, that's out of scope :)

File: Core/Steps/SubSteps/AddFixVersionsForNewReleaseBranchSubStep.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    {
+      AddFixVersionToIssues(rcVersion.ToString(), issues);
+    }
+
+    Console.WriteLine($"Do you want to create the fix version '{nextJiraVersion}' and add it to the issues?");
+    if (InputReader.ReadConfirmation())
+    {
+      AddFixVersionToIssues(nextJiraVersion.ToString(), issues);
+    }
+
```
Comment (MichaelKetting) [inline Core/Steps/SubSteps/AddFixVersionsForNewReleaseBranchSubStep.cs]: extra whitespace found

File: Core/Steps/SubSteps/AddFixVersionsForNewReleaseBranchSubStep.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    PrintIssueListing(issues);
+
+    var rcVersion = currentVersion.GetNextRC();
+    Console.WriteLine($"Do you want to create the fix version '{rcVersion}' and add it to the issues?");
+    if (InputReader.ReadConfirmation())
+    {
+      AddFixVersionToIssues(rcVersion.ToString(), issues);
+    }
+
+    Console.WriteLine($"Do you want to create the fix version '{nextJiraVersion}' and add it to the issues?");
```
Comment (MichaelKetting) [inline Core/Steps/SubSteps/AddFixVersionsForNewReleaseBranchSubStep.cs]: wording:
> Do you want to create version '{nextJiraVersion}' in JIRA and add it as fix version for to the issues?

(also two lines before. maybe the wording is not perfect yet?)

File: Core/Steps/SubSteps/ReleaseVersionAndMoveIssuesSubStep.cs [hunk 9]
```diff
@@ -66,99 +75,67 @@ public void Execute (SemanticVersion currentVersion, SemanticVersion nextVersion
...
+    {
+      issuesToMove = Array.Empty<JiraToBeMovedIssue>();
+      return false;
+    }
+
+    issuesToMove = JiraIssueService.FindAllNonClosedIssues(versionID);
+    if (issuesToMove.Count == 0)
+      return false;
+
+    Console.WriteLine("These are some of the issues that will be moved by releasing the version on jira:");
```
Comment (MichaelKetting) [inline Core/Steps/SubSteps/ReleaseVersionAndMoveIssuesSubStep.cs]:
```suggestion
    Console.WriteLine("These are some of the issues that will be moved by releasing version '{versionname}' on JIRA:");
```

File: Core/Steps/SubSteps/ReleaseVersionAndMoveIssuesSubStep.cs [hunk 10]
```diff
@@ -66,99 +75,67 @@ public void Execute (SemanticVersion currentVersion, SemanticVersion nextVersion
...
+    PrintIssueListing(issuesToMove);
+    Console.WriteLine("Do you want to move these issues to the new version and release the old one or just release the old version?");
```
Comment (MichaelKetting) [inline Core/Steps/SubSteps/ReleaseVersionAndMoveIssuesSubStep.cs]: what does YES and No mean, here. please reword or rebuild to multiple choice. Don't know which is better in the workflow without demo

File: Core/Steps/SubSteps/ReleaseVersionAndMoveIssuesSubStep.cs [hunk 11]
```diff
@@ -66,99 +75,67 @@ public void Execute (SemanticVersion currentVersion, SemanticVersion nextVersion
...
       {
         var message = $"Could not find any issues only associated with fixVersion '{currentFullVersion}', no moving of these issues is necessary.";
-        _console.WriteLine(message);
+        Console.WriteLine(message);
         _log.Debug(message);
         return;
       }
 
-      _console.WriteLine($"The following Jira tickets are closed and only associated with version '{currentFullVersion}':");
+      Console.WriteLine($"The following Jira tickets are closed and only associated with version '{currentFullVersion}':");
```
Comment (MichaelKetting) [inline Core/Steps/SubSteps/ReleaseVersionAndMoveIssuesSubStep.cs]: We got some more jira/Jira casings. do you want to fix them?

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
I dodn't check the tests in depth, am sure @a-ctor checked if it looks good. got some text bits to improve.

File: Core/Steps/SubSteps/ReleaseVersionAndMoveIssuesSubStep.cs (hunk 11 above)
Comment (Frazerus) [reply Core/Steps/SubSteps/ReleaseVersionAndMoveIssuesSubStep.cs]: Fixed the ones I found that are close to the already done changes, might need a cleanup task

**Review by MichaelKetting — APPROVED** [review]

**Pull request merged**